feat: [Geneva Exporter] Add span upload support to Geneva exporter

lalitb opened the pull request:
Adds OpenTelemetry span upload functionality to the Geneva exporter, enabling trace data ingestion alongside existing log support.

__Key Changes:__

- __geneva-uploader__: Added span encoding with Bond format, single payload approach using "Span" event name for routing
- __opentelemetry-exporter-geneva__: New `GenevaTraceExporter` implementing `SpanExporter` trait with concurrent upload support
- __Dependencies__: Added "trace" feature to opentelemetry-proto for OTLP span data structures
- __Example__: Added `trace_basic.rs` demonstrating span creation and export with SimpleSpanProcessor
- JSON serialization for span links (C# compatibility)

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

utpilla commented on opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs:
```diff
@@ -245,6 +367,95 @@ impl OtlpEncoder {
+        use std::hash::{Hash, Hasher};
+
+        // Pre-allocate with estimated capacity to avoid reallocations
+        let estimated_capacity = 12 + span.attributes.len(); // 7 always + 5 max conditional + attributes
```
This calculation seems wrong. It looks like you have missed the three Part A conditional fields.

utpilla commented on opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs:
```diff
@@ -259,6 +470,116 @@ impl OtlpEncoder {
+
+        // Pre-calculate timestamps
+        let start_timestamp = Self::format_timestamp(span.start_time_unix_nano);
+        let formatted_timestamp = start_timestamp.clone(); // Use start time as primary timestamp
```
I don't think this clone is needed.

utpilla commented on opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs:
```diff
@@ -259,6 +470,116 @@ impl OtlpEncoder {
+                }
+                _ => {
+                    // Handle dynamic attributes
+                    if let Some(attr) = span.attributes.iter().find(|a| a.key == field.name) {
```
Do we need to check if it's present in span attributes? Every item in `&[FieldDef]` input should be valid and serialized, right?

utpilla commented on opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs:
```diff
@@ -674,4 +995,163 @@ mod tests {
+        let metadata = "namespace=testNamespace/eventVersion=Ver1v0";
+        let result = encoder.encode_span_batch([span].iter(), metadata).unwrap();
+
+        assert!(!result.is_empty());
```
Use a stricter assertion.

```suggestion
        assert_eq!(result.len(), 1);
```

lalitb commented:
Thanks for the review, @utpilla! I’ve addressed all the comments. I’ll go ahead and merge this — any additional feedback can be handled in a follow-up PR.
